src/plugins/intel_gpu/src/graph/impls/onednn/primitive_onednn_base.h

```diff
@@ -44,11 +47,64 @@ struct typed_primitive_onednn_impl : public typed_primitive_impl<PType> {
           _outer(arg),
           _desc(desc),
           _attrs(attrs),
-          _pd(pd),
-          _prim(pd) { }
+          _pd(pd) {
+            build_primitive();
+        }
 
     bool is_cpu() const override { return false; }
 
+private:
+    std::string get_cache_directory() const {
+        auto path = _outer.get_program().get_engine().configuration().kernels_cache_path;
+        if (path.empty()) {
+            return {};
+        }
+
+        if (path.back() != '/' && path.back() != '\\') {
+            path += "/";
+        }
+        return path;
+    }
+
+    std::string generate_cache_path_from_key(std::vector<uint8_t> key) const {
+        auto path = get_cache_directory();
+        if (path.empty()) {
+            return {};
+        }
+
+        std::string key_str(key.begin(), key.end());
+        size_t hash = std::hash<std::string>()(key_str);
+        return path + std::to_string(hash) + ".onednn.cl_cache";
+    }
+
+    void build_primitive() {
+        auto cache_outpath = get_cache_directory();
+        if (cache_outpath.empty()) {
+            _prim = PrimType(_pd);
+        } else {
+            std::vector<uint8_t> key = _pd.get_cache_blob_id();
+            assert(!key.empty());
+
+            std::vector<uint8_t> cache;
+            {
+                std::lock_guard<std::mutex> lock(cacheAccessMutex);
```

Comment:
Have you checked the performance impact from this mutex? I'm just curious if we can have some load time reduction by using std::shared_mutex from c++17

Reply:
Not formally, but did some informal testing at the beginning of this task. Actually, no performance impact was observed.
However, there might be some perf overhead(e.g. when HDD is used or it is not cached in Linux kernel's cache), and I think it is correct to use such rw lock, instead of vanilla mutex. Let's follow-up on that matter separately.